src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,738 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
-			this->AddAction(STR_NETWORK_CLIENTLIST_SPEAK_TO_COMPANY, &ClientList_SpeakToCompany);
+		/* Disable the chat button when there are players in this company. */
+		chat_button->disabled = !has_players;
+	}
+
+	/**
+	 * Rebuild the list, meaning: calculate the lines needed and what buttons go on which line.
+	 */
+	void RebuildList()
+	{
+		NetworkClientInfo *own_ci = NetworkClientInfo::GetByClientID(_network_own_client_id);
+
+		this->buttons.clear();
+		this->line_count = 0;
+		this->player_host_index = -1;
+		this->player_self_index = -1;
+
+		/* As spectator, show a line to create a new company. */
+		if (own_ci->client_playas == COMPANY_SPECTATOR && !NetworkMaxCompaniesReached()) {
+			this->buttons[line_count].emplace_back(new CompanyButton(SPR_JOIN, STR_NETWORK_CLIENT_LIST_NEW_COMPANY_TOOLTIP, COLOUR_ORANGE, COMPANY_SPECTATOR, &NetworkClientListWindow::OnClickCompanyNew));
+			this->line_count += 1;
 		}
-		this->AddAction(STR_NETWORK_CLIENTLIST_SPEAK_TO_ALL, &ClientList_SpeakToAll);
 
-		/* A server can kick clients (but not himself). */
-		if (_network_server && _network_own_client_id != ci->client_id) {
-			this->AddAction(STR_NETWORK_CLIENTLIST_KICK, &ClientList_Kick);
-			this->AddAction(STR_NETWORK_CLIENTLIST_BAN, &ClientList_Ban);
+		if (own_ci->client_playas != COMPANY_SPECTATOR) {
+			this->RebuildListCompany(own_ci->client_playas, own_ci);
+		}
+
+		/* Companies */
+		for (const Company *c : Company::Iterate()) {
+			if (c->index == own_ci->client_playas) continue;
+
+			this->RebuildListCompany(c->index, own_ci);
+		}
+
+		/* Spectators */
+		this->RebuildListCompany(COMPANY_SPECTATOR, own_ci);
+
+		this->vscroll->SetCount(this->line_count);
+	}
+
+	ButtonCommon *GetButtonAtPoint(Point pt)
+	{
+		uint index = this->vscroll->GetScrolledRowFromWidget(pt.y, this, WID_CL_MATRIX);
+		NWidgetBase *widget_matrix = this->GetWidget<NWidgetBase>(WID_CL_MATRIX);
+
+		bool rtl = _current_text_dir == TD_RTL;
+		uint x = rtl ? (uint)widget_matrix->pos_x + WD_FRAMERECT_LEFT : widget_matrix->current_x - WD_FRAMERECT_RIGHT;
+
+		/* Find the buttons for this row. */
+		auto button_find = this->buttons.find(index);
+		if (button_find == this->buttons.end()) return nullptr;
+
+		/* Check if we want to display a tooltip for any of the buttons. */
+		for (auto &button : button_find->second) {
+			uint left = rtl ? x : x - button->width;
+			uint right = rtl ? x + button->width : x;
+
+			if (IsInsideMM(pt.x, left, right)) {
+				return button.get();
+			}
+
+			int width = button->width + WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT;
+			x += rtl ? width : -width;
 		}
 
-		this->InitNested(client_id);
-		CLRBITS(this->flags, WF_WHITE_BORDER);
+		return nullptr;
 	}
 
-	Point OnInitialPosition(int16 sm_width, int16 sm_height, int window_number) override
+public:
+	NetworkClientListWindow(WindowDesc *desc, WindowNumber window_number) :
+			Window(desc),
+			hover_index(-1),
+			player_self_index(-1),
+			player_host_index(-1)
```

Comment:
Don't think we have a consistent codestyle for constructor lists, but I'm pretty sure it's not this :)

```suggestion
	NetworkClientListWindow(WindowDesc *desc, WindowNumber window_number)
			: Window(desc), hover_index(-1), player_self_index(-1), player_host_index(-1)
```

would be better IMO (and make the diff shorter!)

Reply:
https://github.com/OpenTTD/OpenTTD/blob/master/src/script/script_instance.cpp#L51

And many more places do this.

Optimizing for a shorter diff is silly :) And personally, I find my way more readable. But it is a mixed bag atm .. :)